File: src/SnapBrief.Core/Exporting/CaptureLabels.cs
using SnapBrief.Core.Models;

namespace SnapBrief.Core.Exporting;

public readonly record struct LabeledAnnotation(string DisplayLabel, AnnotationItem Annotation);

public static class CaptureLabels
{
    public static string ForIndex(int zeroBasedIndex)
    {
        if (zeroBasedIndex is < 0 or >= 26)
        {
            throw new ArgumentOutOfRangeException(nameof(zeroBasedIndex));
        }

        return ((char)('A' + zeroBasedIndex)).ToString();
    }

    public static string ForAnnotation(string captureLabel, int oneBasedIndex) =>
        $"{captureLabel}{oneBasedIndex}";

    public static IEnumerable<LabeledAnnotation> ForNotedAnnotations(string captureLabel, CaptureItem capture)
    {
        var noteNumber = 0;
        foreach (var annotation in capture.Annotations)
        {
            if (!ExportText.HasContent(annotation.Note))
            {
                continue;
            }

            noteNumber++;
            yield return new LabeledAnnotation(ForAnnotation(captureLabel, noteNumber), annotation);
        }
    }
}

File: macos/Sources/SnapBriefCore/Exporting/CaptureLabels.swift
import Foundation

/// Port of `src/SnapBrief.Core/Exporting/CaptureLabels.cs`.
public struct LabeledAnnotation: Equatable, Sendable {
    public let displayLabel: String
    public let annotation: AnnotationItem
}

public enum CaptureLabels {
    /// Port of `ForIndex` (SPEC-DELTA-2B §B): 0 -> "A", 1 -> "B", ..., 25 -> "Z", 26 -> "AA", ...
    /// (bijective base-26). Errors only when `zeroBasedIndex < 0`; unlike the pre-sync-2 version,
    /// there is no upper bound.
    public static func forIndex(_ zeroBasedIndex: Int) throws -> String {
        guard zeroBasedIndex >= 0 else {
            throw SnapBriefError.argumentOutOfRange("zeroBasedIndex")
        }
        var value = zeroBasedIndex + 1
        var label = ""
        while value > 0 {
            value -= 1
            let scalar = Unicode.Scalar(UInt8(65 + value % 26))
            label = String(Character(scalar)) + label
            value /= 26
        }
        return label
    }

    /// Port of `ForAnnotation`, e.g. ("A", 1) -> "A1".
    public static func forAnnotation(captureLabel: String, oneBasedIndex: Int) -> String {
        "\(captureLabel)\(oneBasedIndex)"
    }

    /// Port of `ForNotedAnnotations`.
    public static func forNotedAnnotations(captureLabel: String, capture: CaptureItem) -> [LabeledAnnotation] {
        var result: [LabeledAnnotation] = []
        var noteNumber = 0
        for annotation in capture.annotations {
            guard ExportText.hasContent(annotation.note) else { continue }
            noteNumber += 1
            result.append(
                LabeledAnnotation(
                    displayLabel: forAnnotation(captureLabel: captureLabel, oneBasedIndex: noteNumber),
                    annotation: annotation))
        }
        return result
    }
}
